handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java

```diff
@@ -461,7 +464,7 @@ public final long sslCtxPointer() {
     //            get access to an OpenSslSessionContext after this method was called to prevent the user from
     //            producing a segfault.
     final void destroy() {
-        synchronized (ReferenceCountedOpenSslContext.class) {
+        if (DESTROYED_UPDATER.compareAndSet(this, 0, 1)) {
```

Comment:
There are methods like `OpenSslServerSessionContext#getSessionCacheSize()` which use `ctx` that may cause issues (e.g. core dump). I know there is a "warning" message above this method but I think we should consider one of the the following approaches:

1. use synchronized to protect access to `ctx`
Just like we protect access to `ssl` in the `OpenSslEngine` we can protect `ctx` so we avoid calling native methods when `ctx == NULL`.
2. ensure native methods all check for `ctx == NULL`
If native methods are "safe" (don't core dump ... throwing an exception is fine) then maybe there is no risk? As long as we set `ctx = 0` before we call `SSLContext.free(ctx)` this should be OK.

If we can get away with approach 2 we can avoid synchronized. Can we use the same approach in `OpenSslEngine` to avoid synchronization?